### ISSUE TITLE
Legacy autograd function with non-static forward method is deprecated and will be removed in 1.3 and UserWarning: Legacy autograd function object was called twice.  You will probably get incorrect gradients from this computation, as the saved tensors from the second invocation will clobber the saved tensors from the first invocation.  Please consider rewriting your autograd function in the modern style; for information on the new format, please see: https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd 626

### ISSUE THREAD

**Issue by Borahb** [pytorch/pytorch#34473]
I have been running this code for object detection but while running the code ,I am getting a warning 

..\torch\csrc\autograd\python_function.cpp:638: UserWarning: Legacy autograd function with non-static forward method is deprecated and will be removed in 1.3. Please use new-style autograd function with static forward method. (Example: https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function)
..\torch\csrc\autograd\python_function.cpp:664: UserWarning: Legacy autograd function object was called twice.  You will probably get incorrect gradients from this computation, as the saved tensors from the second invocation will clobber the saved tensors from the first invocation.  Please consider rewriting your autograd function in the modern style; for information on the new format, please see: https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd
626

after every step.

## Code

#Importing of Libraries
import torch
from torch.autograd import Variable
import cv2
from data import BaseTransform, VOC_CLASSES as labelmap
from ssd import build_ssd
import imageio

# Defining a function that will do the detection
def detect(frame, net, transform):
    height, width = frame.shape[:2]
    frame_t = transform(frame)[0]
    x = torch.from_numpy(frame_t).permute(2,0,1)
    x = Variable(x.unsqueeze(0))
    y = net(x)
    detections = y.data # We create the detections tensor contained in the output y.
    scale = torch.Tensor([width, height, width, height]) # We create a tensor object of dimensions [width, height, width, height].
    for i in range(detections.size(1)): # For every class:
        j = 0 # We initialize the loop variable j that will correspond to the occurrences of the class.
        while detections[0, i, j, 0] >= 0.6: # We take into account all the occurrences j of the class i that have a matching score larger than 0.6.
            pt = (detections[0, i, j, 1:] * scale).numpy() # We get the coordinates of the points at the upper left and the lower right of the detector rectangle.
            cv2.rectangle(frame, (int(pt[0]), int(pt[1])), (int(pt[2]), int(pt[3])), (255, 0, 0), 2) # We draw a rectangle around the detected object.
            cv2.putText(frame, labelmap[i - 1], (int(pt[0]), int(pt[1])), cv2.FONT_HERSHEY_SIMPLEX, 2, (255, 255, 255), 2, cv2.LINE_AA) # We put the label of the class right above the rectangle.
            j += 1 # We increment j to get to the next occurrence.
    return frame

# Creating the SSD neural network
net = build_ssd('test')
net.load_state_dict(torch.load('ssd300_mAP_77.43_v2.pth', map_location = lambda storage, loc: storage ))

# Creating the transformation
transform = BaseTransform(net.size, (104/256.0, 117/256.0, 123/256.0))
 

# Doing some object detection on a video
reader = imageio.get_reader('ghystreet.mp4')
fps = reader.get_meta_data()['fps']
writer = imageio.get_writer('ghystreetout.mp4', fps = fps)
for i, frame in enumerate(reader): # We iterate on the frames of the output video:
    frame = detect(frame, net.eval(), transform) # We call our detect function (defined above) to detect the object on the frame.
    writer.append_data(frame) # We add the next frame in the output video.
    print(i) # We print the number of the processed frame.
writer.close() # We close the process that handles the creation of the output video.

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by peterjc123** [comment]
Please prepend the following code to ignore warnings.
```
import warnings
warnings.filterwarnings("ignore")
```

**Comment by Borahb** [comment]
okay...but for the  2nd warning it is also affecting the gradients,its saying to rewrite the autograd function in new style. So where to change ?

**Comment by peterjc123** [comment]
Please use `Tensor` instead of `Variable`. More details could be retrieved here. https://pytorch.org/blog/pytorch-0_4_0-migration-guide/#merging-tensor-and-variable-and-classes

**Comment by Borahb** [comment]
I did but it could not import Tensor, i watched the blog, but couldn't identify what to change.

**Comment by ngimel** [comment]
You have to rewrite your custom autograd functions with @staticmethod forward and backward methods, as the warning suggests.

**Comment by Borahb** [comment]
Can you specify on which file the changes have to be made.

**Comment by albanD** [comment]
In your net, you use a custom autograd.Function that is an old style one.  As mentionned in the warning, you should follow the instructions in this doc on how to write a proper custom Function: https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function

You have to make the change wherever you define the new `autograd.Function`.

Note that this warning will become an error in 1.5 as mentionned in the message.

**Comment by turabihasan** [comment]
i have the same problem. did you solve it?

Legacy autograd function with non-static forward method is deprecated. Please use new-style autograd function with static forward method. (Example: https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function)

**Comment by albanD** [comment]
Hi,

As the error mentions you should define the Function as in the doc. You can find more information [here](https://pytorch.org/docs/stable/notes/extending.html) as well.
